src/_posts/2017-02-09-stateful-processing.md

```diff
@@ -0,0 +1,524 @@
+gaps, and every element gets an index.
+
+Conceptually expressing this as a stateful loop is as trivial as you can
+imagine: The state you should store is the next index.
+
+ - As an element comes in, output it along with the next index.
+ - Increment the index.
+
+This presents a good opportunity to talk about big data and parallelism,
+because the algorithm in those bullet points is not parallelizable at all! If
+you wanted to apply this logic over an entire `PCollection`, you would have to
+process each element of the `PCollection` one-at-a-time... this is obvious a
+bad idea.  State in Beam is tightly scoped so that most of the time a stateful
+`ParDo` transform should still be possible for a runner to execute in parallel,
+though you still have to be thoughtful about it.
+
+A state cell in Beam is scoped to a key+window pair. When your DoFn reads or
+writes state by the name of `"index"`, it is actually accessing a mutable cell
+specified by `"index"` _along with_ the key and window currently being
+processed.  So, when thinking about a state cell, it may be helpful to consider
+the full state of your transform as a table, where the rows are named according
+to names you use in your program, like `"index"`, and the columns are
+key+window pairs, like this:
+
+|               | (key, window)<sub>1</sub> | (key, window)<sub>2</sub> | (key, window)<sub>3</sub> | ... |
+|---------------|---------------------------|---------------------------|---------------------------|-----|
+| `"index"`       | `3`                         | `7`                         | `15`                        | ... |
+| `"fizzOrBuzz?"` | `"fizz"`                    | `"7"`                       | `"fizzbuzz"`                | ... |
+| ...           | ...                       | ...                       | ...                       | ... |
+{:.table}
+
+(if you have a superb spatial sense, feel free to imagine this as a cube where
+keys and windows are independent dimensions)
+
+You can provide the opportunity for parallelism by making sure that table has
+enough columns, either via many keys in few windows - for example, a globally
+windowed stateful computation keyed by user ID - or via many windows over few
+keys - for example, a fixed windowed stateful computation over a global key.
+Caveat: all Beam runners today parallelize only over the key.
+
+Most often your mental model of state can be focused on only a single column of
+the table, a single key+window pair. Cross-column interactions do not occur
+directly, by design.
+
+## State in Beam's Java SDK
+
+Now that I have talked a bit about stateful processing in the Beam model and
+worked through an abstract example, I'd like to show you what it looks like to
+write stateful processing code using Beam's Java SDK.  Here is the code for a
+stateful `DoFn` that assigns an arbitrary-but-consistent index to each element
+on a per key-and-window basis:
+
+```java
```

Comment:
I spent ~5 minutes in the Javascript to hack something but didn't get it working. I added Python placeholders so it wasn't mysteriously absent, and the user will have the call to action to switch to Java.